### DISCUSSION TITLE
Command line tool for bulk action on multiple repositories

### DISCUSSION THREAD

**Discussion by luolong** [GitoxideLabs/gitoxide#478] [Ideas]
I am sure I am not the only one around who has needed something along these lines, but almost every time I need to work on some project or other, I am faced with having to juggle more than one git repository at a time.

Often this means checking out N repositories, making a change in one or more of these, pushing a change to remotes, keeping local clones up-to date, rebasing, mergeing, etc. You get the gist.

To do that I've written couple of versions of a tool that would allow me to do common operations on a bunch of repos with a single command. My latest attempt [is now up on GitHub](https://github.com/luolong/git-all) so that I would not have to write it again next time I switch my laptop. Don't mind the code as much as the general idea of it.

Basically, I want a tool that would behave mostly as a regular `git` command line, but would perform it's actions on all git repos within current working directory.

In addition it would have few commands render somewhat different output where it makes sense.
I have implemented table formatted version of `git-status` for example that displays a table of statuses details for each repository it found.

What I am getting at is a suggestion to implement such a tool inside **_gitoxide_** itself.
That could bring some significant performance and user experience benefits as it would be much easier to parallelize all operations and have a more dynamic output compared to a shell scripting language.

What do you think?

**Comment by Byron** [comment]
Indeed I feel that problem even though ultimately I only work on a single repository at a time.

That said, `gitoxide` provides a very generic solution in the form of `ein t find`, which will find all repositories within the CWD or given directory and print their paths line by line. From there it's possible to build all kinds of things within ones shell.

The functionality is also [available as library](https://github.com/Byron/gitoxide/blob/6f4eea936d64fb9827277c160f989168e7b1dba2/gitoxide-core/src/organize.rs#L218) for consumption within Rust programs.

`gitoxide` is also a platform for applications, and as such it should provide the building blocks to do what you are suggestion and at this point I believe it already does. Once it's sufficiently clear how a `multi` tool could look like there is no reason not to add it to `gitoxide` either, but for now I'd hope folks could experiment to figure out what a good baseline for such multi-tool should be.
